core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala

```diff
@@ -219,11 +219,25 @@ class ReplayListenerSuite extends SparkFunSuite with BeforeAndAfter with LocalSp
     // Verify the same events are replayed in the same order
     assert(sc.eventLogger.isDefined)
     val originalEvents = sc.eventLogger.get.loggedEvents
+      .map(JsonProtocol.sparkEventFromJson(_))
     val replayedEvents = eventMonster.loggedEvents
-    originalEvents.zip(replayedEvents).foreach { case (e1, e2) =>
-      // Don't compare the JSON here because accumulators in StageInfo may be out of order
-      JsonProtocolSuite.assertEquals(
-        JsonProtocol.sparkEventFromJson(e1), JsonProtocol.sparkEventFromJson(e2))
+      .map(JsonProtocol.sparkEventFromJson(_))
+    // Executor metrics updates are not logged, so do not get replayed.
+    // Stage executor metrics are logged at stage completion, for any of the executors and
+    // the driver for which we have metrics. We always have metrics for executors, because
+    // they are sent at task end as well as in executor metrics updates. We do not always
+    // have metrics for the driver, because they are only sent in executor metrics updates
+    // (at heartbeat intervals), and when we do, it is only in the original events, never
+    // in the replay, since executor metrics updates are not replayed.
+    // For this reason, exclude stage executor metrics for the driver.
+    val filteredEvents = originalEvents.filter { e =>
+      if (e.isInstanceOf[SparkListenerStageExecutorMetrics] &&
+        e.asInstanceOf[SparkListenerStageExecutorMetrics].execId == "driver") {
```

Comment:
What I am not understanding is that, why are both different? They make JSONs from the same instances. Are they different? Why is it flaky?

Reply:
EventMonster extends EventLoggingListener which doesn't transparently log events, hence some possible cases could exist where both can be different.

Reply:
I tried to explain briefly in the comment. Let me try to explain in more detail.

When the application is run, executor metrics update events may occur (if the application runs long enough). These events are not written to the event log, so they do not get replayed. That is the root cause of differences.
On executor metrics update, the EventLoggingListener updates a map it uses to track per-stage metrics, but does not log. The listener also receives metrics on task end. Because of this, it always has metrics for executors. On stage completion. the listener logs stage executor metrics, for any of the executors/driver it has metrics for. As explained, sometimes it will have metrics for the driver (if an executor metrics update arrived from the driver), but most of the time not.
To recap, SparkListenerStageExecutorMetrics events are not events received while the application is running, but secondary events derived from metrics received and written to the event log. (One other point not relevant to this bug but may help understanding - on replay, the SparkListenerStageExecutorMetrics events do get replayed but are ignored by the listener.)

Reply:
I had written the same explanation in https://github.com/apache/spark/pull/25659#issuecomment-527741025.